src/purge.rs

```diff
@@ -323,32 +332,56 @@ where
                 m.region_id()
             };
 
-            // FIXME: This code makes my brain hurt.
-            let mut cursor = 0;
-            while cursor < entry_indexes.len() {
-                let entry =
-                    read_entry_bytes_from_file(self.pipe_log.as_ref(), &entry_indexes[cursor])?;
-                total_size += entry.len();
-                entries.push(entry);
-                if total_size > MAX_REWRITE_BATCH_BYTES {
-                    let mut take_entries = Vec::with_capacity(expect_rewrites_per_memtable);
-                    std::mem::swap(&mut take_entries, &mut entries);
-                    let mut take_entry_indexes = entry_indexes.split_off(cursor + 1);
-                    std::mem::swap(&mut take_entry_indexes, &mut entry_indexes);
-                    log_batch.add_raw_entries(region_id, take_entry_indexes, take_entries)?;
-                    self.rewrite_impl(&mut log_batch, rewrite, false)?;
-                    total_size = 0;
-                    cursor = 0;
-                } else {
-                    cursor += 1;
+            // Split the entries into smaller chunks, so that we don't OOM, and the
```

Comment:
Can following logic be implemented as two functions depending on whether `atomic_group` is used? I find it quite hard to understand.

Reply:
Because use_atomic_group=true doesn't necessarily mean atomic_group.is_some(), most of the logic cannot be simplified for the use_atomic_group=true case.

And for the atomic_group.is_some() case, I think the logic is relatively simple (only involves one branch).

Reply:
I tried the second time, still found the code hard to read and lost in context switch.

Reply:
🤷 I have refactored it several times too, I guess it's just that complex.